.github/workflows/ci.yml

```diff
@@ -0,0 +1,108 @@
+name: ci
+
+on:
+  push:
+    branches: [main]
+  pull_request:
+    branches: [main]
+
+env:
+  CARGO_TERM_COLOR: always
+  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
+  CONTAINER_REGISTRY: ghcr.io
+  CONTAINER_IMAGE_NAME: ${{ github.repository }}
+
+jobs:
+  build_and_test:
+    if: ( ! github.event.pull_request.draft )
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v3
+
+      - name: Install stable rust toolchain
+        uses: dtolnay/rust-toolchain@stable
+        with:
+          toolchain: stable
+
+      - name: Use caching
+        uses: Swatinem/rust-cache@v2.5.0
+
+      - name: "Build"
+        run: cargo build
+
+      - name: "Test"
+        run: RUST_LOG=DEBUG cargo test --bins --tests --examples --workspace -- --nocapture
+
+  rustfmt:
+    name: Rustfmt
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v3
+
+      - name: Install nightly rust toolchain with rustfmt
+        uses: dtolnay/rust-toolchain@stable
+        with:
+          toolchain: nightly
+          components: rustfmt
+
+      - name: "Check formatting"
+        run: cargo +nightly fmt --check --all
+      
+  clippy:
+    name: Clippy
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v3
+
+      - name: Install stable rust toolchain
+        uses: dtolnay/rust-toolchain@stable
+        with:
+          toolchain: stable
+          components: clippy
+
+      - name: "Clippy check"
+        run: cargo clippy --all-features -- -D warnings
+
+  build-and-push-image:
```

Comment:
I was also thinking of that as that is the standard used everywhere else! Though I thought we could keep things simpler by just publishing main's without doing proper versioning first since the server is still in early development phase — or do you think we can actually start doing proper versioning now itself?

Reply:
Will be doing proper versioning